Fix painless casting bug causing opensearch to crash

shiv0408 opened the pull request:
### Description
#### Update
Throwing ClassCastException when trying to cast def to void.
Created a Unit Test to ensure that def to void casting throws ClassCastException. Removed the following mention Unit Test as it was just added to reproduce the issue. The current unit test should be enough to avoid any regression in future.

Created a unit test to reproduce the bug #6435 by creating a unit test. Fix for this is requires a small change in  `modules/lang-painless/src/main/java/org/opensearch/painless/AnalyzerCaster.java` to throw `ClassCastException` when we are trying to cast `def` to `void`. But to create this unit test, I had to make change supported scripts in `_scripts/painless/_execute` API by adding the update context. How can we keep this unit test in this code without modifying the supported scripts? One option is we can keep a setting which users can't modify and set it from the test, using that setting allow the update operation in supported scripts.

### Related Issues
Resolves #6435 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented:
> I had to make change supported scripts in _scripts/painless/_execute API by adding the update context

So this is a new feature? I don't know this code well, what does this do? What are the downsides? If it's new, you should make a PR adding the feature first, then another one that fixes the bug we're talking about on top, just to keep things clean.

shiv0408 commented:
> So this is a new feature?

This change would allow users to run painless scripts with update context. Although, it is not my intention to add this as a new feature. Rather than creating a new feature, I wanted to keep context which is only used in tests and not by customers.

> what does this do? What are the downsides?

This [API](https://opensearch.org/docs/latest/api-reference/script-apis/exec-script/) allows customers to execute painless scripts with are not stored. Providing an Update context script has no major use case here.

reta commented:
@shiv0408 could you please clarify what is supposed outcome of this script (from the issue):

```
curl -X POST "localhost:9200/hockey/_update/3?pretty" -H 'Content-Type: application/json' -d'
{
  "script": {
    "source": "def x=1;return x;",
    "lang": "painless",
    "params": {
      "sold_cost": 26
    }
  }
}
```

- the  `sold_cost` is not used - why it is there?
- the `ctx._source` is not used - nothing to update?

shiv0408 commented:
@reta I was modifying some script to reproduce the issue. The purpose of this script is just to demonstrate the issue, so did not use `ctx._source` to update any doc and `param` is also not required, left it by mistake.

shiv0408 commented:
And I am calling this update context, but any void context like `ingest` or `reindex` can face same issue if we try to return def type in a void expecting context.

reta commented:
> And I am calling this update context, but any void context like `ingest` or `reindex` can face same issue if we try to return def type in a void expecting context.

So it the context is void expected - why script should return anything? And what is happening with the results that script returns? (as I see here, we just drop it)

shiv0408 commented:
Ideally, script should not return anything. But, if we do return something it should just throw exception or handle it gracefully and not crash the OpenSearch process.

shiv0408 commented:
@reta @dblock I have added the code which would throw the error when trying to cast from def to void type. But the UT I added in `PainlessExecuteApiTests.java` also passes now, but do we want to keep this test going ahead? I have also added a UT to ensure that we throw exception when trying to return def type from void context.

reta commented on modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -577,6 +579,13 @@ static Response innerShardOperation(Request request, ScriptService scriptService
                     double result = scoreScript.execute(null);
                     return new Response(result);
                 }, indexService);
+            } else if (scriptContext == UpdateScript.CONTEXT) {
```
We don't need the `UpdateScript.CONTEXT` right?

shiv0408 replied:
We don't need it to fix the bug. I added it to reproduce the issue in a Unit Test.

shiv0408 commented:
These changes are not necessary to fix the bug. I added them to repro the issue in a UT. As I have added a separate UT in `FactoryTests` we can remove this test.

shiv0408 commented:
Thanks for approval @reta. Added DCO on all commits.

reta commented:
@dblock looks good to you?

shiv0408 commented:
@reta Can we try to get this merged before 2.9 branch cut?

reta commented:
> @reta Can we try to get this merged before 2.9 branch cut?

@dblock had comments / concerns, we need his approval

dblock commented on modules/lang-painless/src/test/java/org/opensearch/painless/FactoryTests.java:
```diff
@@ -358,6 +358,11 @@ public void testVoidReturn() {
             () -> getEngine().compile("void_return_test", "1 + 1", VoidReturnTestScript.CONTEXT, Collections.emptyMap())
         );
         assertEquals(iae.getMessage(), "not a statement: result not used from addition operation [+]");
+        ClassCastException exception = expectScriptThrows(
```
Make this test separate?

shiv0408 replied:
Created a different set in new commit

shiv0408 commented:
@dblock I think it should be good to merge now.
